### PR TITLE
Add `no-arguments-for-html-elements` rule

### PR DESCRIPTION
**Pull request by lifeart**
ref: https://github.com/ember-template-lint/ember-template-lint/issues/926
//cc @jenweber, @rwjblue

### CODE REVIEW

**Comment by lifeart** [comment]
@rwjblue should we split "checks" to different rules or figure out new rule name to cover all cases? 
Or just remove some "checks"..

**Comment by Alonski** [comment]
@lifeart Can the name of this PR be changed to something like:
`Feature: No arguments for html elements`?

File: test/unit/rules/lint-no-arguments-for-html-elements-test.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    <:slot @name="Header"></:slot>
+  </MyComponent>`,
+    '<@foo.bar @name="2" />',
+    '<@foo @name="2" />',
+    '<foo.some.name @name="1"/>',
+  ],
+
+  bad: [
+    {
+      template: '<div as |a|></div>',
```
Comment (Turbo87) [inline test/unit/rules/lint-no-arguments-for-html-elements-test.js]: this looks unrelated to a rule called `no-arguments-for-html-elements`. IMHO we should extract that into a `no-block-params-for-html-elements` rule.

File: test/unit/rules/lint-no-arguments-for-html-elements-test.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      result: {
+        message: ERROR_MESSAGE_3.replace('%', 'a'),
+        moduleId: 'layout.hbs',
+        source: '<div as |a|></div>',
+        line: 1,
+        column: 0,
+      },
+    },
+    {
+      template: '<div name=@value></div>',
```
Comment (Turbo87) [inline test/unit/rules/lint-no-arguments-for-html-elements-test.js]: same as above, this should probably be extracted to a dedicated rule, because

1. it would be useful for component invocations too
2. there might be cases where you actually want to pass a `@`-prefixed string as the attribute and in that case it should be possible to disable the rule

Comment (lifeart) [reply test/unit/rules/lint-no-arguments-for-html-elements-test.js]: any suggestions how this roule could be named? (require-mustache-wrap-variables) ?

 for prefixed string it should be `name="@" or name = '@' ` and we can check it using `sourceForNode` method + indexOf

File: docs/rule/no-arguments-for-html-elements.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
+## no-arguments-for-html-elements
```
Comment (bmish) [inline docs/rule/no-arguments-for-html-elements.md]: Can you mention how "angle bracket components" use the `@argument` style and that's why users are more likely to start making this mistake?

There's also a rule documentation template FYI: https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/_TEMPLATE_.md

File: lib/rules/lint-no-arguments-for-html-elements.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
+'use strict';
+
+const Rule = require('./base');
+
+const ERROR_MESSAGE = 'Argument "%" won\'t work for HTML elements ("<%>").';
```
Comment (bmish) [inline lib/rules/lint-no-arguments-for-html-elements.js]: Instead of creating this custom string interpolation scheme with the percent signs, you should just use and export functions that look like:

```
function makeErrorMessage(argumentName) {
  return `...${argumetName}`;
}
```

File: test/unit/rules/lint-no-arguments-for-html-elements-test.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
+'use strict';
+
+const generateRuleTests = require('../../helpers/rule-test-harness');
+const ERROR_MESSAGE = require('../../../lib/rules/lint-no-arguments-for-html-elements')
+  .ERROR_MESSAGE;
+const ERROR_MESSAGE_2 = require('../../../lib/rules/lint-no-arguments-for-html-elements')
+  .ERROR_MESSAGE_2;
+const ERROR_MESSAGE_3 = require('../../../lib/rules/lint-no-arguments-for-html-elements')
```
Comment (bmish) [inline test/unit/rules/lint-no-arguments-for-html-elements-test.js]: No need to require the same file several times. Just do:
```
const rule = require(...);
const { ERROR_MESSAGE_1, ERROR_MESSAGE_2, ... } = rule; 
```

File: test/unit/rules/lint-no-arguments-for-html-elements-test.js (hunk 2 above)
Comment (Turbo87) [reply test/unit/rules/lint-no-arguments-for-html-elements-test.js]: maybe `no-at-prefixed-string-values` or something like that

I'm not sure I understand the second question 🤔

Comment (lifeart) [reply test/unit/rules/lint-no-arguments-for-html-elements-test.js]: 2.) solved using this approach -> https://github.com/ember-template-lint/ember-template-lint/pull/927/files#diff-407e0d43d58a271ba8e91f1876ea034fR37

File: docs/rule/no-arguments-for-html-elements.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    at CompilableBlock.compile
+    at Object.evaluate
+    at AppendOpcodes.evaluate
+    at LowLevelVM.evaluateSyscall
+    at LowLevelVM.evaluateInner
+    at LowLevelVM.evaluateOuter
+    at VM.next
+```
+
+This is creepy and untracable (for end-users).
```
Comment (MelSumner) [inline docs/rule/no-arguments-for-html-elements.md]:
```suggestion
This is untraceable for end-users.
```

File: docs/rule/no-arguments-for-html-elements.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+Using this linting rule we can guess (is it component or not) and emit error if argument assigned to lowercased element without same block name and without dots in notation.
```
Comment (MelSumner) [inline docs/rule/no-arguments-for-html-elements.md]:
```suggestion
This linting rule determines if it is a component or not and emits an error if argument is assigned to a lower-case element without the same block name and without dots in notation.
```

File: docs/rule/no-arguments-for-html-elements.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+This rule also emit error if we trying to specify "path-like" string as attribute argument.
```
Comment (MelSumner) [inline docs/rule/no-arguments-for-html-elements.md]:
```suggestion
This rule will also emit an error for these other conditions: 

- trying to specify "path-like" string as attribute argument
- trying to use an HTML element with block params

```

File: docs/rule/no-arguments-for-html-elements.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+Also, error will be thrown if we trying to use HTML element with block-params.
```
Comment (MelSumner) [inline docs/rule/no-arguments-for-html-elements.md]:
```suggestion
```

Added this line to the bullet points above

File: docs/rule/no-arguments-for-html-elements.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+```hbs
+<foo.some.name @name="1"/>
+<@foo @name="2" />
+<@foo.bar @name="2" />
+```
+
+```hbs
+<MyComponent>
+  <:slot @name="Header"></:slot>
```
Comment (MelSumner) [inline docs/rule/no-arguments-for-html-elements.md]: is this legit valid code?

File: lib/rules/lint-no-arguments-for-html-elements.js [hunk 11]
```diff
@@ -0,0 +1,103 @@
+'use strict';
+
+const Rule = require('./base');
+
+const ERROR_MESSAGES = {
+  1: 'Argument "%" won\'t work for HTML elements ("<%>").',
+  2: 'Attribute value "%" won\'t work as expected, you need to wrap it as "%".',
+  3: 'Block param "%" won\'t work for HTML element.',
```
Comment (MelSumner) [inline lib/rules/lint-no-arguments-for-html-elements.js]:
```suggestion
  3: 'Block params ("%") should not be used on HTML elements.',
```

File: lib/rules/lint-no-arguments-for-html-elements.js [hunk 12]
```diff
@@ -0,0 +1,103 @@
+'use strict';
+
+const Rule = require('./base');
+
+const ERROR_MESSAGES = {
+  1: 'Argument "%" won\'t work for HTML elements ("<%>").',
```
Comment (MelSumner) [inline lib/rules/lint-no-arguments-for-html-elements.js]:
```suggestion
  1: 'Arguments ("%") should not be used on HTML elements ("<%>").',
```

File: lib/rules/lint-no-arguments-for-html-elements.js [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  2: 'Attribute value "%" won\'t work as expected, you need to wrap it as "%".',
```
Comment (MelSumner) [inline lib/rules/lint-no-arguments-for-html-elements.js]: I'm not sure what you're trying to say here?

File: docs/rule/no-arguments-for-html-elements.md [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+</MyComponent>
+```
```
Comment (MelSumner) [inline docs/rule/no-arguments-for-html-elements.md]: You're missing some of the template.md sections here- having a references section is important at the very least.

Comment (lifeart) [reply docs/rule/no-arguments-for-html-elements.md]: @MelSumner could you mention some references? It's removed because I can't figure out matched refs.

File: docs/rule/no-arguments-for-html-elements.md (hunk 10 above)
Comment (lifeart) [reply docs/rule/no-arguments-for-html-elements.md]: yes, this is "slots" (already available in glimmer-vm)

File: lib/rules/lint-no-arguments-for-html-elements.js (hunk 13 above)
Comment (lifeart) [reply lib/rules/lint-no-arguments-for-html-elements.js]: @MelSumner in handlebars AST this code is equal:

`<div name=@foo />` and `<div name="@foo" />`, but first one have no sence (likely an error), and second one is "string param @foo"

Comment (lifeart) [reply lib/rules/lint-no-arguments-for-html-elements.js]: https://github.com/ember-template-lint/ember-template-lint/pull/927#discussion_r349895000

File: lib/rules/no-arguments-for-html-elements.js [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+      }, []);
+      if (!safeValues.includes(originalNode)) {
+        log({
+          message: makeError(2, `${value.chars}`, `${name}={{${value.chars}}}`),
+          line: value.loc && value.loc.start.line,
+          column: value.loc && value.loc.start.column,
+          source: sourceForNode(value),
+        });
+      }
+    }
```
Comment (rwjblue) [inline lib/rules/no-arguments-for-html-elements.js:49]: I don't fully understand the point of this set of checks, can you explain?

Note, we already have [no-quoteless-attributes](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/no-quoteless-attributes.md) in the recommended rule set.

File: lib/rules/no-arguments-for-html-elements.js [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+    const sourceForNode = this.sourceForNode.bind(this);
+
+    function looksLikeHTMLElement(node) {
+      if (isLocal(node)) {
+        return false;
+      }
+      const tagName = node.tag;
+      if (tagName.startsWith('@') || tagName.startsWith(':') || tagName.includes('.')) {
+        return false;
+      }
```
[the remark this answers is not in the document]
Comment (lifeart) [reply lib/rules/no-arguments-for-html-elements.js:29]: is angle-baracket component return false for `<foo.some.name />`

File: lib/rules/no-arguments-for-html-elements.js (hunk 15 above)
Comment (lifeart) [reply lib/rules/no-arguments-for-html-elements.js]: good catch! removed

File: docs/rule/no-arguments-for-html-elements.md (hunk 10 above)
Comment (fusion2004) [reply docs/rule/no-arguments-for-html-elements.md:99]: :O Is there documentation on how to use or examples of this? Or is this not recommended to use in production apps yet?

File: test/unit/rules/lint-no-arguments-for-html-elements-test.js (hunk 1 above)
Comment (Turbo87) [reply test/unit/rules/lint-no-arguments-for-html-elements-test.js]: gotta say I'm surprised that this got merged without splitting it into several rules, especially now that it's planned to be included in the recommended ruleset...

Comment (rwjblue) [reply test/unit/rules/lint-no-arguments-for-html-elements-test.js]: hmmm, ya I must have missed this. Good catch, we should absolutely split it out.

Comment (rwjblue) [reply test/unit/rules/lint-no-arguments-for-html-elements-test.js]: Made https://github.com/ember-template-lint/ember-template-lint/issues/1451 to track, and added to v3.0.0 milestone.
